### PR TITLE
Add support for STM32 ADC clock and prescaler

### LINKED ISSUE #40955: ADC: STM32: Add device tree config for clock source
Looking here https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/adc_stm32.c#L897, you can see that ADC clock source is hard-coded to `LL_ADC_CLOCK_ASYNC_DIV4`. I'd like to see some config added to the device tree to allow one to specify this.

I really need some help on what would be a good device tree addition to represent this.

### PR DESCRIPTION
**Pull request by gautierg-st**
This PR adds the support and configuration of the full clock capabilities of the STM32 ADCs.

On many series, the ADC can be configured to use as clock source either a synchronous clock or an asynchronous one, though only one of them may be available on some series. Furthermore, a prescaler is configurable to divide the chosen clock. Also, depending on the series, the asynchronous clock can be selected between different clocks (for example SYSCLK, PLL out, HSE...).

Functionalities vary between series, and this PR tries to address all cases, while remaining as minimal as possible.
- This PR first deals with the clocks of the F1 / F3 series that are fully / partly configured in RCC. New RCC bindings are thus added for them to allow configuring the ADC prescaler clock. Split in #60332
- The support for the HSI14 dedicated ADC clock is added for the F0 family. Split in #60333
- For all series except F1 and F37x, two new ADC bindings are added to define the type of clock, and the prescaler. When combined, these values are used to configure the clock in the ADC registers.
- Finally, the support for secondary/kernel source clock is added, allowing to configure the source of the asynchronous clock. This works just the same as the secondary clock for other STM32 drivers (like UART, SPI or I2C)

Fixes #40955

### CODE REVIEW

File: dts/bindings/clock/st,stm32f1-rcc.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2021, Linaro ltd
```
Comment (erwango) [inline dts/bindings/clock/st,stm32f1-rcc.yaml:1]: Please fix copyrights

File: dts/arm/st/f3/stm32f373.dtsi [hunk 2]
```diff
@@ -11,6 +11,11 @@
 	soc {
 		compatible = "st,stm32f373", "st,stm32f3", "simple-bus";
 
+		rcc: rcc@40021000 {
+			/* ADC prescaler is the same as on STM32F1 */
```
Comment (erwango) [inline dts/arm/st/f3/stm32f373.dtsi:15]: This comment could be confusing (for someone not aware of the global PR context).

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/clock/st,stm32f1-rcc.yaml (hunk 1 above)
Comment (gautierg-st) [reply dts/bindings/clock/st,stm32f1-rcc.yaml:1]: Done in #60332

File: dts/arm/st/f3/stm32f373.dtsi (hunk 2 above)
Comment (gautierg-st) [reply dts/arm/st/f3/stm32f373.dtsi:15]: Done in #60332

**Review by erwango — CHANGES_REQUESTED** [review]
Some comments, mainly about bindings

File: dts/bindings/adc/st,stm32-adc.yaml [hunk 3]
```diff
@@ -27,6 +27,44 @@ properties:
   pinctrl-names:
     required: true
 
+  st,adc-clock-source:
+    type: int
+    enum:
+      - 1 # SYNC for synchronous ADC clock source
+      - 2 # ASYNC for asynchronous ADC clock source
+    default: 1
```
[the remark this answers is not in the document]
Comment (gautierg-st) [reply dts/bindings/adc/st,stm32-adc.yaml:35]: The default values of SYNC-4 work on most series. Only L1, U5 and WBA have no synchronous clock. For them, I defined ASYNC directly in the dtsi (And ASYNC-4 is available).
This allows a working configuration out of the box, but it is completely arbitrary.
If we don't set defaults, then we need to set the properties required, and we need to define them in all boards dts that use ADC.

[the remark this answers is not in the document]
Comment (erwango) [reply dts/bindings/adc/st,stm32-adc.yaml:35]: ok, what is the default register value ?

[the remark this answers is not in the document]
Comment (gautierg-st) [reply dts/bindings/adc/st,stm32-adc.yaml:35]: It depends on series, but overall this is not the default hardware register value. So I guess we remove the default and set it for each board

[the remark this answers is not in the document]
Comment (erwango) [reply dts/bindings/adc/st,stm32-adc.yaml:35]: Unfortunately, yes, I think this is what needs to be done.

**Comment by erwango** [comment]
Ok for me, aside from CI issues

**Review by erwango — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**